### ISSUE TITLE
New feature - live migrate mailboxes with imapsync

### ISSUE THREAD

**Issue by chris001** [webmin/webmin#595]
imapsync is an very reliable free open source perl app which syncs/migrates email mailbox accounts.

It'd be great to add this as a feature, to eliminate the need to run a snapshot backup and restore of many hundreds of GB of mailbox data.

Installing the app is very easy, I write a script to install it with all of its requirements.

Syntax to call the command is very easy. Parameters are source and destination mail server names, account login names, passwords, plus some flag options for dry run, sync internal dates, and which header is the unique key ('Message-Id').

As it's a very long running process, it'd be necessary to call it through `screen` or `tee` so that it would run in its own process and survive restarts of the webmin `miniserv.pl` process.  Virtualmin could reconnect to it to display current status of the sync process.

Then obviously you'll want to run it a 2nd and 3rd and 4th time to sync mails which have arrived during the previous syncs so it's all caught up, before you cut over the DNS to using your new mails server as the primary or only MX.  Which you reduced the TTL on the MX to 30-60 sec, before performing the initial imapsync, so that when you do the cut over, the update to MX propagates quickly.

**Comment by jcameron** [comment]
Isn't imapsync for syncing folders between different machines using the IMAP protocol?

A better solution in Virtualmin for more incremental backups would be `rsync` . Also, we already support incremental backups using `tar`.

**Comment by chris001** [comment]
Yes `imapsync` is for syncing mailboxes between 2 remote IMAP servers.

On the virtualmin side, you have access to the filesystem holding the mailboxes.

On the remote side, you often don't have access to the filesystem holding the mailboxes (gmail. yahoo, hotmail, fastmail, gmx, any commercial mail host).

The only universal way to migrate IMAP mailboxes to/from virtualmin is `imapsync`.

**Comment by swelljoe** [comment]
There's already an IMAP client in Usermin (maybe in Webmin, too, since they share a lot of mail code). You wouldn't need to use a third party tool for grabbing mail from an IMAP server; it could just be done with the existing IMAP client functions.

**Comment by swelljoe** [comment]
What's the use case you're wanting to solve with this feature?

**Comment by chris001** [comment]
Use case: Have 160 GB of mail in about 20 mailboxes, located on remote linux centos 6 server running dovecot.  No shell access.  Migrate that mail out of that remote server to virtualmin ubuntu 16 with full shell access.  Because of the huge amount of data, the only reliable way, with least chance of losing any mail, to get all the mail out of the remote centos no shell, and into the local virtualmin, is via IMAP. `imapsync` is one of the best apps for this.

**Comment by gnadelwartz** [comment]
its good to know about imap sync, thanks for this.
on the other hand it seems to be very specialised and webmin has already tools for similar use cases.

do you have a suggestion what existing module should make use of imap sync?

**Comment by chris001** [comment]
@gnadelwartz  Which existing module to extend to use with imapsync?  Possibly the "Edit Users" screen/module. This is where each user in the domain is displayed, along with their email mailbox size in use on disk.  There could be a button added, Import Remote Mailboxes.
For efficiency and good user experience (or admin experience), you should let the user specify a CSV type file, with one line per remote + local email account to sync.

The file.txt file contains:
```
remote_server1_1:user001_1;password001_1;local_server1_2:user001_2;password001_2
remote_server2_1:user002_1;password002_1;local_server2_1:user002_2;password002_2
remote_server3_1:user003_1;password003_1;local_server3_2:user003_2;password003_2
remote_server4_1:user004_1;password004_1;local_server4_2:user004_2;password004_2
remote_server5_1:user005_1;password005_1;local_server5_2:user005_2;password005_2
```

Each mailbox sync will probably be a long running process, most likely between 2 minutes and possibly more than 24 hours for massive 100GB mailboxes on to distant, high-roundtrip-latency servers.  So you need spawn a `screen` and then let the user click on other links in virtualmin, while that `screen` console runs in the background doing the `imapsync`.  When the virtualmin user comes back to this page to view the status of the sync, they should have the ability to get a live look of the `screen` and/or a summary of the most recent X messages from the `screen` which will be `imapsync` printing lines to the `screen` console such as, `message 1089 sync OK, size 82KB, messages remaining 98772`.

**Comment by gnadelwartz** [comment]
Live Migration of a mailbox, espacially from a not owned server, has many pitfalls.
independent from used tool,  I would never force my endusers to hand out their password to migrate anything.

As I already sometimes migrated from an old to a new server, I see two posibilities to do this:

1.  snyc mailboxes with rsync. But this needs admin acess to both servers, the servers needs to be the same type and needs to be handeled carefull if both servers can recieve new mail
2.  setup an fechmail job per enduser for to import mail from old server. User has to fill in his password when starting migration.

imapsync can do this job also, but I would do this anly form my own accounts from command line.

even I'm not in the position to judge about webmin (or more paticular virtualmin) features, I would use a fetchmail based solution per enduser mailbox to import/migrate external mail to webmin

**Comment by chris001** [comment]
* `imapsync` does the same job as `rsync` when you don't have admin shell access to the servers, you just have server name, username, and password.  
* `imapsync` preserves the email flags (e.g. you already read the message), while `fetchmail` ignores the flags and makes all the downloaded emails appear as new and unread, correct?

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
